clang/runtime/dpct-rt/include/dpl_extras/algorithm.h

```diff
@@ -1746,6 +1746,91 @@ segmented_reduce_argmax(Policy &&policy, Iter1 keys_in, Iter2 keys_out,
+
+  auto n = ::std::distance(input_beg, input_end);
+  // First element must be nontrivial run (start of first segment)
+  auto first_adj_it = oneapi::dpl::adjacent_find(policy, input_beg, input_end);
+  auto first_adj_idx = ::std::distance(input_beg, first_adj_it);
+  if (first_adj_it == input_end) {
+    ::std::fill(policy, num_runs, num_runs + 1, 0);
+    return ::std::make_pair(offsets_out, lengths_out);
+  }
+  auto get_prev_idx_element = [first_adj_idx](const auto &idx) {
+    auto out_idx = idx + first_adj_idx;
+    return (out_idx == 0) ? 0 : out_idx - 1;
+  };
+  auto get_next_idx_element = [first_adj_idx, n](const auto &idx) {
+    auto out_idx = idx + first_adj_idx;
+    return (out_idx == n - 1) ? n - 1 : out_idx + 1;
+  };
+  // TODO: Use shifted view to pad range once oneDPL ranges is non-experimental
+  auto left_shifted_input_beg =
+      oneapi::dpl::make_permutation_iterator(input_beg, get_prev_idx_element);
+  auto right_shifted_input_beg =
+      oneapi::dpl::make_permutation_iterator(input_beg, get_next_idx_element);
+  // Segment type for ith idx consists of zip of iterators at (i-1, i, i+1)
+  // padded at the ends
+  auto zipped_keys_beg = make_zip_iterator(
+      left_shifted_input_beg, input_beg, right_shifted_input_beg,
+      oneapi::dpl::counting_iterator<offsets_t>(0));
+  // Set flag at the beginning of new nontrivial run (ex: (2, 3, 3) -> 1)
+  auto key_flags_beg =
+      make_transform_iterator(zipped_keys_beg, [n](const auto &zipped) {
+        using ::std::get;
+        bool last_idx_mask = get<3>(zipped) != n - 1;
+        return (get<0>(zipped) != get<1>(zipped) &&
+                get<1>(zipped) == get<2>(zipped)) &&
+               last_idx_mask;
+      });
+  auto count_beg = oneapi::dpl::counting_iterator<offsets_t>(0);
+  auto const_it = dpct::make_constant_iterator(lengths_t(1));
+  // Check for presence of nontrivial element at current index
+  auto tr_nontrivial_flags = make_transform_iterator(
+      make_zip_iterator(left_shifted_input_beg, input_beg),
+      [](const auto &zip) {
+        using ::std::get;
+        return get<0>(zip) == get<1>(zip);
+      });
+  auto zipped_vals_beg =
+      make_zip_iterator(tr_nontrivial_flags, count_beg, const_it);
+  auto pred = [](bool lhs, bool rhs) { return !rhs; };
+  // The first call to op in reduce_by_segment will have the second of the input
+  // values as the rhs parameter as the first is passed as lhs. The first
+  // element's flag of 1 due to the padded left shift will always be ignored and
+  // will never be added to the length count which is only updated in
+  // get<2>(lhs) += get<0>(rhs) and initialized to 1.
+  auto op = [](auto lhs, const auto &rhs) {
+    using ::std::get;
+    // Update count of run. The run's starting index is stored in get<1>(lhs) as
+    // the initial value in the segment and is preserved throughout the
+    // segment's reduction.
+    get<2>(lhs) += get<0>(rhs);
+    return ::std::move(lhs);
+  };
```

Comment:
```suggestion
  auto op = [](auto lhs, const auto &rhs) {
    using ::std::get;
    
    // Update length count of run.
    // The first call to this op will use the first element of the input as lhs and second element
    // as rhs.  get<0>(first_element) is always ignored in favor of a constant `1` in get<2>, 
    // avoiding the need for special casing the first element. The constant `1` utilizes the 
    // knowledge that each segment begins with a non-trivial run.
    get<2>(lhs) += get<0>(rhs);
    
    // The run's starting index is stored in get<1>(lhs) as
    // the initial value in the segment and is preserved throughout the
    // segment's reduction to end up as the non-trivial run's offset.
    
    return ::std::move(lhs);
  };
```

I took a stab at revising these comments to makes things a little more clear.  Take/adapt them if you wish.

Reply:
Thanks, I like this revisions. I have added them with some minor wording changes.